File: Coding_Questions/Python/RunLengthEncoding.py
# ---Run Length Encoding---


def run_length_encoding(string: str) -> str:
    encoded_string = ""
    previous_char = string[0]
    counter = 0

    for char in string:
        if char == previous_char:
            counter += 1

            if counter > 9:
                encoded_string += f"9{previous_char}"
                counter = 1

            continue
        encoded_string += str(counter) + previous_char
        previous_char = char
        counter = 1

    encoded_string += str(counter) + previous_char

    return encoded_string


input_string = "AAAAAAAAAAAAABBCCCCDD"
expected = "9A4A2B4C2D"
actual = run_length_encoding(input_string)
print(actual)
assert actual == expected

File: Coding_Questions/C++/RunLengthEncoding.cpp
// ---Run-Length Encoding---


std::string run_length_encoding(std::string str) {
    std::string encoding;
    char previous_char = str[0];
    size_t counter = 1;

    for (size_t i = 1; i < str.length(); ++i) {
        if (str[i] == previous_char) {
            ++counter;

            if (counter > 9) {
                encoding += std::to_string(9) + previous_char;
                counter = 1;
            }
        } else {
            encoding += std::to_string(counter) + previous_char;
            previous_char = str[i];
            counter = 1;
        }
    }

    encoding += std::to_string(counter) + previous_char;

    return encoding;
}